### PR TITLE
Sdw rework

### PR DESCRIPTION
**Pull request by RanderWang**
In this PR,  some sdw special operations are moved to sdw dai driver, such as prepare, trigger and stream free, so that we don't need to check dai index or get sdw stream data form hda data.

### CODE REVIEW

File: sound/soc/codecs/rt700.c [hunk 1]
```diff
@@ -1269,24 +1263,17 @@ static int rt700_pcm_hw_params(struct snd_pcm_substream *substream,
...
-
-	retval = sdw_stream_add_slave(rt700->slave, &stream_config,
-				      &port_config, 1, stream->sdw_stream);
-	if (retval) {
-		dev_err(dai->dev, "Unable to configure port\n");
-		return retval;
-	}
+	port_config = &rt700->port_config;
+	port_config->ch_mask = (1 << (num_channels)) - 1;
+	port_config->num = port;
```
Comment (plbossart) [inline sound/soc/codecs/rt700.c:1276]: port is not initialized?

Comment (plbossart) [inline sound/soc/codecs/rt700.c:1276]: Can you explain why you need those two structures? Is this to memorize information that will be used in prepare? If yes, maybe you can add comments to help the reviewer, there are none in your changes...

File: sound/soc/codecs/rt700.c [hunk 2]
```diff
@@ -1339,7 +1326,33 @@ static int rt700_pcm_hw_params(struct snd_pcm_substream *substream,
...
+	stream = snd_soc_dai_get_dma_data(dai, substream);
+
+	if (!stream) {
+		dev_err(dai->dev, "no dma data found in dai %s", dai->name);
+		return -ENOMEM;
+	}
+
+	retval = sdw_stream_add_slave(rt700->slave, &rt700->stream_config,
+				      &rt700->port_config, 1,
+				      stream->sdw_stream);
```
Comment (plbossart) [inline sound/soc/codecs/rt700.c:1349]: I don't fully get this code, is this going to initialize the 'stream'?
Again please add comments to explain what you are doing, the code is not self-explanatory.

File: drivers/soundwire/intel.c [hunk 3]
```diff
@@ -593,6 +593,17 @@ static int intel_config_stream(struct sdw_intel *sdw,
...
+			     struct snd_pcm_substream *substream,
+			     struct snd_soc_dai *dai)
+{
+	if (sdw->res->ops && sdw->res->ops->free_stream && sdw->res->arg)
+		return sdw->res->ops->free_stream(sdw->res->arg,
+				substream, dai);
+
+	return 0;
+}
+
```
Comment (plbossart) [inline drivers/soundwire/intel.c:606]: this should be separate patch

File: drivers/soundwire/intel.c [hunk 4]
```diff
@@ -738,6 +749,62 @@ static int intel_startup(struct snd_pcm_substream *substream,
...
+			return ret;
+		}
+	}
+
+	return 0;
+
+error:
+	kfree(name);
+	sdw_release_stream(sdw_stream);
+	return ret;
```
Comment (plbossart) [inline drivers/soundwire/intel.c:805]: this is not done in the right order, you need to revert the kfree and sdw_release_stream

File: drivers/soundwire/intel.c [hunk 5]
```diff
@@ -738,6 +749,62 @@ static int intel_startup(struct snd_pcm_substream *substream,
...
+	else
+		snprintf(name, 32, "%s-Capture", dai->name);
+
+	sdw_stream = sdw_alloc_stream(name);
+	if (!sdw_stream) {
+		dev_err(dai->dev,
+			"alloc stream failed for DAI %s: %d",
+			dai->name, ret);
+		ret = -ENOMEM;
+		goto error;
```
Comment (plbossart) [inline drivers/soundwire/intel.c:775]: you only kfree(name) here

File: drivers/soundwire/intel.c [hunk 6]
```diff
@@ -738,6 +749,62 @@ static int intel_startup(struct snd_pcm_substream *substream,
...
+	}
+
+	/* Set stream pointer on CPU DAI */
+	ret = snd_soc_dai_set_sdw_stream(dai,
+					 sdw_stream,
+					 substream->stream);
+	if (ret < 0) {
+		dev_err(dai->dev, "failed to set stream pointer on cpu dai %s",
+			dai->name);
+		return ret;
```
Comment (plbossart) [inline drivers/soundwire/intel.c:785]: memory leak, use goto error

**Review by plbossart — CHANGES_REQUESTED** [review]
Thanks @RanderWang this looks very interesting.
I didn't fully get the code for the RT700, can you please comment more, and the error handling needs to be improved, but this would be good if we can avoid changing the SOF parts.
Nice work. 
btw when you re-submit this can you add it on top of sof/integration/soundwire?

File: drivers/soundwire/intel.c [hunk 7]
```diff
@@ -738,6 +749,62 @@ static int intel_startup(struct snd_pcm_substream *substream,
...
+			"alloc stream failed for DAI %s: %d",
+			dai->name, ret);
+		ret = -ENOMEM;
+		goto error;
+	}
+
+	/* Set stream pointer on CPU DAI */
+	ret = snd_soc_dai_set_sdw_stream(dai,
+					 sdw_stream,
+					 substream->stream);
```
Comment (ranj063) [inline drivers/soundwire/intel.c:781]: @RanderWang please use one line if possible.

File: drivers/soundwire/intel.c [hunk 8]
```diff
@@ -738,6 +749,62 @@ static int intel_startup(struct snd_pcm_substream *substream,
 	return ret;
 }
 
+static int sdw_stream_setup(struct snd_pcm_substream *substream,
+			    struct snd_soc_dai *dai)
+{
+	struct snd_soc_pcm_runtime *rtd = substream->private_data;
+	struct sdw_stream_runtime *sdw_stream = NULL;
+	char *name;
+	int i, ret = 0;
```
Comment (ranj063) [inline drivers/soundwire/intel.c:758]: superfluous initialisation?

File: drivers/soundwire/intel.c [hunk 9]
```diff
@@ -738,6 +749,62 @@ static int intel_startup(struct snd_pcm_substream *substream,
...
+
+	/* Set stream pointer on all CODEC DAIs */
+	for (i = 0; i < rtd->num_codecs; i++) {
+		ret = snd_soc_dai_set_sdw_stream(rtd->codec_dais[i],
+						 sdw_stream,
+						 substream->stream);
+		if (ret < 0) {
+			dev_err(dai->dev, "failed to set stream pointer on codec dai %s",
+				rtd->codec_dais[i]->name);
+			return ret;
```
Comment (ranj063) [inline drivers/soundwire/intel.c:796]: memory leak here too?

File: drivers/soundwire/intel.c [hunk 10]
```diff
@@ -750,6 +817,12 @@ static int intel_hw_params(struct snd_pcm_substream *substream,
 	int ret, i, ch, dir;
 	bool pcm = true;
 
+	ret = sdw_stream_setup(substream, dai);
```
Comment (ranj063) [inline drivers/soundwire/intel.c:820]: Why do we do sdw_stream_step() in hw_params().I think it should be done in startup() to keep it in line with what is there already

File: sound/soc/codecs/rt700.c (hunk 2 above)
Comment (ranj063) [reply sound/soc/codecs/rt700.c:1349]: @RanderWang yeah same here. I dont understand why we need to move this to prepare. AFAICT, sdw_stream_add_slave() need the stream_config to be set, particularly the frame_rate.

File: sound/soc/codecs/rt700.c [hunk 11]
```diff
@@ -1339,7 +1326,33 @@ static int rt700_pcm_hw_params(struct snd_pcm_substream *substream,
...
+	if (retval) {
+		dev_err(dai->dev, "Unable to configure port\n");
+		return retval;
```
Comment (ranj063) [inline sound/soc/codecs/rt700.c:1352]: @RanderWang just return retval outside the if. No need for 2 separate return statements here

**Comment by ranj063** [comment]
@RanderWang the second and third commits need to be squashed

File: sound/soc/sof/intel/hda.c [hunk 12]
```diff
@@ -81,8 +81,36 @@ static int sdw_config_stream(void *arg, void *s, void *dai,
...
+		dev_err(sdev->dev, "error: failed to set DAI hw_free for ALH %d\n",
+			config.dai_index);
+	}
+
+	return ret;
+}
+
 static const struct sdw_intel_ops sdw_callback = {
 	.config_stream = sdw_config_stream,
+	.free_stream = sdw_free_stream,
```
Comment (ranj063) [inline sound/soc/sof/intel/hda.c:113]: @RanderWang where will this get called? Would it be in the hw_free dai op

File: drivers/soundwire/intel.c (hunk 5 above)
Comment (plbossart) [reply drivers/soundwire/intel.c:775]: Actually this is wrong. The concept of 'stream' means we can use multiple DAIs (cpu and codec), so doing a sdw_alloc_stream here is just not quite right. see comments

````c
/**
 * sdw_alloc_stream() - Allocate and return stream runtime
 *
 * @stream_name: SoundWire stream name
 *
 * Allocates a SoundWire stream runtime instance.
 * sdw_alloc_stream should be called only once per stream. Typically
 * invoked from ALSA/ASoC machine/platform driver.
 */
````

File: drivers/soundwire/intel.c (hunk 10 above)
Comment (plbossart) [reply drivers/soundwire/intel.c:820]: I don't think it should be done in a dai operation, the notion of stream is larger than a dai.
For example the same stereo 'stream' can be rendered by using link0-dai0 and link1-dai4, going to two separate amplifiers. This setup should be done in a dailink operation where we will have the list of all cpu and codec dais handled.

Comment (plbossart) [reply drivers/soundwire/intel.c:820]: I am still struggling here, in the initial proposal there was no notion of DAI, 

Initially we did all the soundwire stream handling in pcm_ops callbacks.

Here everything is done as part of dai_ops callbacks.

That's very very different!

Comment (ranj063) [reply drivers/soundwire/intel.c:820]: @plbossart but all the dai_ops are called from the pcm ops context anyway isnt it? like for example: 
https://github.com/thesofproject/linux/blob/b45315d57e2ab6fce1de8bd6391d925143b4fbda/sound/soc/soc-pcm.c#L806

Comment (plbossart) [reply drivers/soundwire/intel.c:820]: Sorry @ranj063, I don't get your point. I never fully understood how DPCM works so not sure if the two models are was referring to are equivalent.

Does anyone have a sequence diagram showing how those rtd, pcm_ops and  cpu_dai and codec_dai ops are called?

File: drivers/soundwire/intel.c (hunk 7 above)
Comment (RanderWang) [reply drivers/soundwire/intel.c:781]: Thanks！I get it.

File: drivers/soundwire/intel.c (hunk 10 above)
Comment (RanderWang) [reply drivers/soundwire/intel.c:820]: I will check the ASOC and give a process flow picture

Comment (bardliao) [reply drivers/soundwire/intel.c:820]: > I don't think it should be done in a dai operation, the notion of stream is larger than a dai.
> For example the same stereo 'stream' can be rendered by using link0-dai0 and link1-dai4, going to two separate amplifiers. This setup should be done in a dailink operation where we will have the list of all cpu and codec dais handled.

I agree with @plbossart the operation can be done in dailink.  Actually I think intel_hw_params() will be invoked twice if there are two sdw components in the same dailink. @RanderWang Can we allocate and release stream in machine driver? Maybe we can use startup and shutdown ops?

Comment (RanderWang) [reply drivers/soundwire/intel.c:820]: Drawing1

Comment (RanderWang) [reply drivers/soundwire/intel.c:820]: There are component pcm driver and dai driver in ASoC.  they all have prepare, hw_params, and trigger functions. In the following picture,   dai hw_params is for dai driver , and  the only hw_params is for component pcm driver. Now there is only component pcm driver in SOF for SDW,  and there is only dai driver in SDW component and codec comopnent.

All the components pcm driver are invoked in both FE and BE dai, so you will find some functions are invoke twice. Although there is no component pcm driver in SDW and codec component, I still marked some functions in in picture to reflect the process flow in ASoC

Drawing2

**Comment by RanderWang** [comment]
@plbossart  @ranj063  Thanks for your review !  I know there are many defects in the PR, I just want to show my idea about SDW.  Now sdw runtime data is stored in dai and shared with sdw dai driver and codec dai driver. I want to move these functions to dai driver to make  SOF  simple.  And also it is easy for CAVS driver to  support SDW, although maybe there is no such requirement.

**Comment by plbossart** [comment]
@RanderWang I did a bit more thinking and I think your idea to implement new DAI callbacks is the right direction. There is one issue with the PR, which I mentioned above, that the sdw_alloc_stream() is not a DAI operation and should be done at the DAI link level. But we can do this later.
The additional simplification would be to remove the callbacks for config and free (using the dai index instead) but again we can do this in an update.
Can you try and provide an updated version on top of sof/integration and we'll work from there?
Thank you.

**Comment by RanderWang** [comment]
> @RanderWang I did a bit more thinking and I think your idea to implement new DAI callbacks is the right direction. There is one issue with the PR, which I mentioned above, that the sdw_alloc_stream() is not a DAI operation and should be done at the DAI link level. But we can do this later.
> The additional simplification would be to remove the callbacks for config and free (using the dai index instead) but again we can do this in an update.
> Can you try and provide an updated version on top of sof/integration and we'll work from there?
> Thank you.

yes, I will refine my PR and submit it.

**Comment by RanderWang** [comment]
done
